go/teams/hidden/ratchet.go

```diff
@@ -0,0 +1,360 @@
+// fits into a RatchetObj; so for instance:
+//
+// [ { b:
+//     { h: <Buffer 57 67 4b a7 c5 ce 55 df 70 cc bf e3 6d 62 4e 56 c2 14 a7 ba 8a fa 84 de d0 70 87 d4 ad 9d 11 e8>,
+//       s: 2, t: 17 },
+//    r:
+//     { k: <Buffer 5f cf 8c 5d 1b 39 2b df 61 e6 94 40 63 3d ea 35 03 8a 91 d1 f4 97 3a fc 59 47 be c4 43 e2 61 e3>,
+//       r: <Buffer 1e 1e 39 42 79 38 aa 0d ff e2 ad c6 32 34 93 f9 ed cb d4 c0 9f 4b 05 b4 b8 84 b0 9e e9 8f d2 b1> },
+//    v: 1 } ]
+//
+// As we can see, r.r corresponds to what was sent in the visible team chain link. r.k is the blinding key.
+// When we compute HMAC-SHA512(r.k, pack(b)), whe should get r.r
+//
+// This file handles encoding/decoding, packing/unpacking, marshalling/unmarshalling of this data.
+//
+
+// EncodedRatchetBlindingKeySet is a b64-encoded, msgpacked map of a RatchetBlindingKeySet, used to POST up to the
+// server 1 new ratchet. Note that even in the case of multiple signatures (inside a TX), it only is really necessary
+// to ratchet the hidden chain once. So this suffices.
+type EncodedRatchetBlindingKeySet string
+
+func (e EncodedRatchetBlindingKeySet) IsNil() bool    { return len(e) == 0 }
+func (e EncodedRatchetBlindingKeySet) String() string { return string(e) }
+
+// BlindingKey is a 32-byte random byte array that is used to blind ratchets, so that they can be
+// selectively hidden via access control.
+type BlindingKey [32]byte
+
+// SCTeamRatchet is the result of HMAC-SHA512(k,v)[0:32], where k is a random Blinding Key,
+// and v is the msgpack of a sig3.Tail.
+type SCTeamRatchet [32]byte
+
+// RatchetVersion is always 1, for now.
+type RatchetVersion int
+
+const RatchetVersion1 = RatchetVersion(1)
+
+type RatchetBlind struct {
+	Hash SCTeamRatchet `codec:"r"`
+	Key  BlindingKey   `codec:"k"`
+}
+
+type RatchetObj struct {
+	Body         sig3.Tail      `codec:"b"`
+	RatchetBlind RatchetBlind   `codec:"r"`
+	Version      RatchetVersion `codec:"v"`
+}
+
+// Ratchet is an object that's used in the teams/teams* and teams/transaction* world to make a visible team chain
+// link incorporate one hidden team ratchet. This means we have to post data both into the signature field (the blinded ratchet)
+// and also data into the sig POST, the blinding keys, etc. This little object conveniniently encapsulates all of that.
+type Ratchet struct {
+	encoded EncodedRatchetBlindingKeySet
+	decoded RatchetObj
+}
+
+// RatchetBlindingKeySet is sent down from the server when we are reading a set of blinding ratchets from
+// the team/get response.
+type RatchetBlindingKeySet struct {
+	m map[SCTeamRatchet]RatchetObj
+}
+
+func (r SCTeamRatchet) String() string {
+	return hex.EncodeToString(r[:])
+}
+
+// UnmarshalJSON is implicitly used in chain_parse.go move SCTeamRatchets into and out of JSON
+// from the hidden team chain.
+func (r *SCTeamRatchet) UnmarshalJSON(b []byte) error {
+	unquoted := keybase1.UnquoteBytes(b)
```

Comment:
Confused about this quoting and unquoting. Are we expecting double encoded JSON here?